dotCMS/src/main/java/com/dotcms/security/secret/SecretsStoreKeyStoreImpl.java

```diff
@@ -0,0 +1,303 @@
+     * 
+     * @return
+     */
+    private File createStoreIfNeeded() {
+        final File secretStoreFile = new File(secretsKeyStorePath);
+        if (!secretStoreFile.exists()) {
+            try {
+                final KeyStore keyStore = KeyStore.getInstance(SECRETS_STORE_KEYSTORE_TYPE);
+                keyStore.load(null, loadStorePassword());
+                flushCache();
+                saveSecretsStore(keyStore);
+            } catch (Exception e) {
+                Logger.warn(this.getClass(), "unable to create secrets store " + SECRETS_STORE_FILE + ": " + e);
+                throw new DotRuntimeException(e);
+            }
+        }
+        return secretStoreFile;
+    }
+
+
+    /**
+     * loads up the Keystore from disk
+     * 
+     * @return
+     */
+    private KeyStore getSecretsStore() {
+        final File secretStoreFile = createStoreIfNeeded();
+        try {
+            final KeyStore keyStore = KeyStore.getInstance(SECRETS_STORE_KEYSTORE_TYPE);
+            try (InputStream in = Files.newInputStream(secretStoreFile.toPath())) {
+                keyStore.load(in, loadStorePassword());
+            }
+
+            return keyStore;
+
+        } catch (Exception e) {
+            Logger.warn(this.getClass(), "unable to load secrets store " + SECRETS_STORE_FILE + ": " + e);
+            throw new DotRuntimeException(e);
+        }
+
+    }
+
+    /**
+     * Persists the keystore, tries to do a 2 phase commit, saving to a tmp file, copying that to the
+     * keystore file, then deleting the tmp file
+     * 
+     * @param keyStore
+     * @return
+     */
+    private KeyStore saveSecretsStore(final KeyStore keyStore) {
+        final File secretStoreFile = new File(secretsKeyStorePath);
+        final File secretStoreFileTmp = new File(secretStoreFile.getParent(), "dotSecretsStore_" + System.currentTimeMillis() + ".p12.tmp");
+
+        secretStoreFileTmp.getParentFile().mkdirs();
+        try (OutputStream fos = Files.newOutputStream(secretStoreFileTmp.toPath())) {
+            keyStore.store(fos, loadStorePassword());
+            FileUtil.copyFile(secretStoreFileTmp, secretStoreFile);
+            flushCache();
+            secretStoreFileTmp.delete();
+        } catch (Exception e) {
+            Logger.warn(this.getClass(), "unable to save secrets store " + secretStoreFileTmp + ": " + e);
+            throw new DotRuntimeException(e);
+        }
+        return keyStore;
+    }
+
+    @VisibleForTesting
+    protected final static char[] CACHE_404 = new char[] {'4', '0', '4'};
+
+    /**
+     * secret values are stored encryped in cache for 30 seconds and are decrypted when requested. If
+     * value is not in cache, it will be read from the keystore
+     * 
+     * @param variableKey
+     * @return
+     * @throws Exception
+     */
+    public Optional<char[]> getValue(final String variableKey) {
+
+        char[] fromCache = getFromCache(variableKey.toLowerCase(),
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Local variable 'fromCache' could be declared final](https://app.codacy.com/app/dotCMS/core/pullRequest?prid=4623512)